In [ ]:
!pip install biopython

from Bio import Entrez, SeqIO
from Bio.SeqUtils import gc_fraction
from collections import Counter

Entrez.email = "rushridhar07@gmail.com"

handle = Entrez.efetch(
    db="nucleotide",
    id="NM_000546",
    rettype="fasta",
    retmode="text"
)

with open("tp53.fasta", "w") as file:
    file.write(handle.read())

handle.close()

record = SeqIO.read("tp53.fasta", "fasta")
sequence = record.seq

length = len(sequence)

A = sequence.count("A")
T = sequence.count("T")
G = sequence.count("G")
C = sequence.count("C")

gc_content = gc_fraction(sequence)*100

print("Sequence ID:", record.id)
print("Total Sequence Length:", length)
print("\nNucleotide Composition:")
print("A:", A)
print("T:", T)
print("G:", G)
print("C:", C)

print(f"\nGC Content: {gc_content:.2f}%")

print("\nFirst 30 nucleotides:")
print(sequence[:30])

print("\nLast 30 nucleotides:")
print(sequence[-30:])

Sequence ID: NM_000546.6
Total Sequence Length: 2512

Nucleotide Composition:
A: 532
T: 639
G: 621
C: 720

GC Content: 53.38%

First 30 nucleotides:
CTCAAAAGTCTAGAGCCACCGTCCAGGGAG

Last 30 nucleotides:
CTCTTATTTTACAATAAAACTTTGCTGCCA


STEPS:
1. Install biopython and import entrez(for ncbi access),SeqIO(for seq parsing) and GC from Bio.SeqUtils(for GC content calculation).
2. The entrez command requires an email to make a request to ncbi database, for fetching the data.
3. Entrez.efetch is used to retrive the TP53 mRNA record record(NM_000546) from NCBI in FASTA format.
4. A new file name is opened "tp53.fasta" in write mode ("w"). Then the handle object which was retrived from the entrez command is parsed using read() and then written in the new file "tp53.fasta".
5. SeqIO.read is used to parse tp53.fasta file and create a SeqRecord object "record". It contains the seq id, name, description and annotations. Whereas, "record.seq" only contains the actual sequence.
6. now i used this record seq variable for diff operations like counting the nucleotides, calculating the gc conetnt.
7. give command to print all the desired outcomes.
8. Result is displayed: Prints the sequence ID, total length, individual nucleotide counts, GC content, and the first and last 30 nucleotides of the sequence.

#Written Analysis
1. The GC content i got after running the code was 53.38% which is higher than 41%. The G and C nucleotide bases form 3 hydrongen bonds instead of 2 and therefore, higher gc conetnt means higher stability of mRNA. The codons that have gc in them are less likely to break down quickly. This results in "codon usage bias". A more stable mRNA can remain in the cell for a longer period of time, hence increasing the expression level of TP53.

2. The first 30 nucleotides of the TP53 mRNA are CTCAAAAGTCTAGAGCCACCGTCCAGGGAG. As discussed earlier, A and T nucleotides form 2 hydrogen bonds which results in les stable secondary structures. The 5' UTR of mRNA being  AT-rich makes it easier for the ribosome to bind and scan it, leading to efficient transalation initiation. In this case the first 30 nucleotides have relatively high gc content especially in the second half of the sequence. Since TP53 is a tumore suppressor gene, this GC content helps with the regulation to ensure that the protein is produced at appropriate levels rather than continuously.

In [ ]:

!pip install biopython

from Bio import Entrez, SeqIO

Entrez.email = "rushridhar07@gmail.com"

handle = Entrez.efetch(
    db="nucleotide",
    id="NM_000546",
    rettype="gb",
    retmode="text"
)

with open("tp53.gb", "w") as file:
    file.write(handle.read())

handle.close()

record = SeqIO.read("tp53.gb", "genbank")

print("Organism:", record.annotations["organism"])

print("\nFirst 4 levels of Taxonomic Lineage:")
lineage = record.annotations["taxonomy"][:4]
for level in lineage:
    print(level)

print("\nMolecule Type:", record.annotations["molecule_type"])
print("Topology:", record.annotations["topology"])

references = record.annotations["references"]
print("\nTotal Number of References:", len(references))

print("\nFirst Reference Title:")
print(references[0].title)

print("\nLast Reference Title:")
print(references[-1].title)


Organism: Homo sapiens

First 4 levels of Taxonomic Lineage:
Eukaryota
Metazoa
Chordata
Craniata

Molecule Type: mRNA
Topology: linear

Total Number of References: 23

First Reference Title:
Therapeutic Editing of the TP53 Gene: Is CRISPR/Cas9 an Option?

Last Reference Title:
Isolation and characterization of a human p53 cDNA clone: expression of the human p53 gene


#STEPS:
1. Install biopython and import entrez(for ncbi access) and SeqIO(for seq parsing).
2. The entrez command requires an email to make a request to ncbi database, for fetching the data.
3. Entrez.efetch is used to retrive the TP53 mRNA record record(NM_000546) from NCBI in GenBank format(specify the format in handle).
4. A new file name is opened "tp53.gb" in write mode ("w"). Then the handle object which was retrived from the entrez command is parsed using read() and then written in the new file "tp53.gb".
5. SeqIO.read is used to parse tp53.fasta file and create a SeqRecord object "record". It contains the seq id, name, description and annotations.
6. Extract and print annotated information using "record.annotations"
7. "organism": Prints the scientific name of the organism from which the sequence originated.
8. "Lineage": Retrieves the full taxonomic lineage and slices it to get only the first four levels as specified in the command.
9. "Molecule Type": Displays the type of molecule (e.g., mRNA, DNA).
10. "Topolgy": Indicates whether the molecule is linear or circular.
11. "References": Retrieves a list of reference objects associated with the record. By using the command "len" it prints the count of these references.
12. "Reference title": Retrives the title attribute of References.
13. Result is displayed: First 4 levels of Taxonomic Lineage, Molecule Type, Topolofy, Count of references and First and last reference titles.


#Written Analysis

1. The taxonomy lineage of TP53 is Eukaryota-> Metazoa-> Chordata-> Craniata. These are bigger groups that include many different organisms, not just humans. Because humans are part of these larger groups, the TP53 gene is not expected to be unique to humans. It is likely that the gene was already present before humans evolved.

2. The last reference is likely the oldest and the first one is the newest.

First Reference Title:
Therapeutic Editing of the TP53 Gene: Is CRISPR/Cas9 an Option?

Last Reference Title:
Isolation and characterization of a human p53 cDNA clone: expression of the human p53 gene

Research pipeline: Discovery → Understanding → Medical application

 In the first study researchers focused on identifying and understanding the gene, while more recent study is exploring how this knowledge can be used to develop therapies for diseases such as cancer.